[WIP] Disable cycling while editing a PM

Subhasish-Behera opened the pull request:
<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->
-> Disables cycling during editing a PM by `tab` key .
->  Shows a warning if one uses tab during editing a PM.
<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->
Partial fix for #774 
<!-- Add a link to a discussion on chat.zulip.org, if relevant -->

**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [X] Manually
- [X] Existing tests (adapted, if necessary)
- [x] New tests added (for any new behavior)
- [X] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->

<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
<!-- For example:
- first commit doing some thing
- maybe multiple commits doing similar things
-->
commit 1: changes made to boxes.py
commit 2: changes made to test_boxes.py
**Notes & Questions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- For example:
- this doesn't include feature X (yet?)
- unsure about Y
- should this do Z?
-->

**Interactions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- eg.
- Waiting on #<PR>
- Blocks #<PR>
-->

**Visual changes** <!-- if any; add/delete/fill-in with screenshot/diagram as appropriate -->
![zt1](https://user-images.githubusercontent.com/92573882/210581649-7b23e518-dd15-4aeb-b3d7-0b2182660ed2.png)

Subhasish-Behera commented:
@zulipbot add "PR needs review"

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -97,6 +97,15 @@ def __init__(self, view: Any) -> None:
         # Prioritizes autocomplete in message body
         self.recipient_user_ids: List[int]
 
+        # Marks if the write box formed is involves [EDIT_MESSAGE: e] key
+        # Only used for private messages(for now)
+        self.edit_keypressed: bool
+        self.edit_keypressed = False
```
The edit state is encoded in `msg_edit_state` - does this give more information?

Subhasish-Behera replied:
I didn't think of that...nope, it does not give any other information. It seems that the same work can be done using `msg_edit_state`(I ran the code replacing `edit_keypressed` with `msg_edit_state` and it worked fine.)
So I will remove the `edit_keypressed` part and use `msg_edit_state`  instead :)

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -97,6 +97,15 @@ def __init__(self, view: Any) -> None:
+        self.edit_keypressed: bool
+        self.edit_keypressed = False
+
+        # Marks if the write box formed is involves [REPLY_MESSAGE: r] key
+        # Only used for private messages(for now)
+        self.reply_keypressed: bool
+        self.reply_keypressed = False
```
Why are we tracking if it's a reply? Ah, this is since private edits are triggered by reusing the reply logic?

Subhasish-Behera replied:
Yes, the keypress of `EDIT_MESSAGE` on private messages works on calling the keypress for `REPLY_MESSAGE` using `self.keypress(size, primary_key_for_command("REPLY_MESSAGE"))`.That's why reply_keypressed is used to differentiate between cases .

neiljp reviewed (commented):
@Subhasish-Behera Thanks for working on this too :+1:

I've tested this manually and it seems to function well, for use of cycling using <kbd>TAB</kbd> or similar.

In the version we merged for streams, it benefits from the stream being a non-editable text object, which also limits clicking into the box or other accidental cases where it receives focus. That is a small issue with the solution in this PR right now, since we have only one private compose setup function.

Taking an approach like for streams but for private messages, seems like it would also avoid the possible need for tracking the reply state. The branch where we reuse the reply mode would then use a dedicated setup function for private edit, like we already have for stream_box_edit_view.

We could still maintain an error message for tab (or up) when editing a PM, as a separate commit/feature, though I don't think it's as important as making the recipient(s) unchangeable.

How does that sound?

Commit:
boxes: Create _setup_common_private_compose.

The new approach is similar to how streams setup the coompose box. Here
_setup_common_private_compose setsup the common requirements of diffrent
write_box view. Mainly the logic behind the self.to_write_box such as
it's auto-complete and when to send typing event updates(yes in reply state
and no in edit state) is not covered here as they are non-common.

Commit:
boxes: Create update_recipients_from_user_ids function.

This function takes in recipient_user_ids and sets the recipient info
accordingly.This function ensures there will be no code repetition. Another
change is making the recipinet_info a class attribute as it will be used
in edit view, non edit view, converting id to info.

Commit:
boxes: Add private_box_edit_view function for edit compose box.

With this function, private messages compose boxes uses 2 different
functions for WriteBox while editing a message and replying a message.
Here the self.to_write_box is non-editable urwid.Text.Also as its not
editable, there is no need for auto-complete(unlike private_box_view).
Its also dosen't send typing event updates to other recipients as it
should be done for new messages and not for editing old messages.

Commit:
messages: Call private_box_edit_view when editing private message.

Earlier editing was dependent on calling the reply_message keypress but
now it calls a separate method custom to creating a WriteBox for editing
private message.

Commit:
tests: boxes: Add tests for _setup_common_private_compose.

Mainly tests if autocomplete has been called, connect_signal of urwid
as msg_write_box is editable and changes are expected.

Commit:
tests: boxes: Add tests for update_recipients_from_user_ids.

The test are based on size of recipient_user_ids list as well
as if it is empty or not.Then the code uses user_id_email_dict and
user_dict to create the end recipient_info.

Commit:
tests: boxes: Add tests for private_box_view.

It tests for autocomplete is called twice, once indirectly for
msg_write_box and other directly for to_write_box.Also asserts
if to_write_box is editable.

Commit:
tests: boxes: Add tests for private_box_edit_view.

It tests if enable_autocomplete is just called once indirectly for
msg_write_box (and not to_write_box).Also checks if to_write_box is
non-editable.

Subhasish-Behera commented:
The new code dosen't need `self.reply_keypressed` or `self.edit_keypressed` to track which key triggered the `private_box_view` as you suggested.

Subhasish-Behera commented:
@zulipbot add "PR needs review"

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -205,18 +205,16 @@ def _setup_common_private_compose(self) -> None:
-        self,
-        *,
-        recipient_user_ids: Optional[List[int]] = None,
+    def convert_id_to_info(
+        self, recipient_user_ids: Optional[List[int]] = None
```
Would this be better incorporated into `_set_regular_and_typing_recipient_user_ids`?

Subhasish-Behera replied:
It would be possible if we were only talking about `convert_id_to_info` but `_set_regular_and_typing_recipient_user_ids` is also used by `update_recipients` function .

neiljp replied:
OK, so
- `_set_regular_and_typing_recipient_user_ids` does the following, dependent upon passed user_ids
  - sets `recipient_user_ids` (autocompletion)
  - sets `typing_recipient_user_ids` (not including user)
- `update_recipients` (which appears to effectively be `update_recipients_from_emails_in_widget` or similar):
  - sets `recipient_emails`
  - calls `_set_regular_and_typing_recipient_user_ids` with calculated ids from recipient_emails
- `convert_id_to_info` does the following, dependent upon the passed user_ids
  - calls `_set_regular_and_typing_recipient_user_ids`
  - sets `recipient_emails`
  - sets `recipient_info`

This pattern suggests that the last two methods have a similar intent and so should share a name prefix, and generally could benefit from renaming.

The latter sets an extra attribute, which seems to be for initialization only - can we make that a return value instead?

neiljp replied:
OK, so
- `_set_regular_and_typing_recipient_user_ids` does the following, dependent upon passed user_ids
  - sets `recipient_user_ids` (autocompletion)
  - sets `typing_recipient_user_ids` (not including user)
- `update_recipients` (which appears to effectively be `update_recipients_from_emails_in_widget` or similar):
  - sets `recipient_emails`
  - calls `_set_regular_and_typing_recipient_user_ids` with calculated ids from recipient_emails
- `convert_id_to_info` does the following, dependent upon the passed user_ids
  - calls `_set_regular_and_typing_recipient_user_ids`
  - sets `recipient_emails`
  - sets `recipient_info`

This pattern suggests that the last two methods have a similar intent and so should share a name prefix, and generally could benefit from renaming.

The latter sets an extra attribute, which seems to be for initialization only - can we make that a return value instead?

neiljp commented on zulipterminal/ui_tools/messages.py:
```diff
@@ -1082,7 +1082,9 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
 
             if self.message["type"] == "private":
-                self.keypress(size, primary_key_for_command("REPLY_MESSAGE"))
+                self.model.controller.view.write_box.private_box_edit_view(
+                    recipient_user_ids=self.recipient_ids,
+                )
```
This is what enables the new edit-private-message functionality, correct?

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -898,22 +898,29 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
-                    # that all the recipients are valid, to avoid including any
-                    # invalid ones.
-                    self.update_recipients(self.to_write_box)
+                    if isinstance(self.to_write_box, ReadlineEdit):
```
Can we use the same conditional here as below? I expect isinstance to work, but it's not as descriptive as the current state.

Subhasish-Behera replied:
If I have understood correctly you are referring to `msg_edit_state`. Yep, that should be a better choice as it seems more basic. Probably used `isinstance(self.to_write_box, ReadlineEdit)` because I was thinking along the lines of `to_write_box` being editable or not as the variable between two cases.

neiljp replied:
Right, `msg_edit_state` is used in the logic below.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -927,6 +927,10 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
             else:
+                if self.msg_edit_state:
+                    self.model.controller.report_error(
+                        [" Recipient(s) can not be edited while editing a PM."]
```
This is now DM, but let's use a fuller version too.

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -1642,7 +1677,10 @@ def test_keypress_CYCLE_COMPOSE_FOCUS(
                 write_box.stream_box_view(stream_id)
         else:
-            write_box.private_box_view()
+            if recipient_box_editable:
+                write_box.private_box_view()
+            else:
+                write_box.private_box_edit_view()
```
We can adjust this test as soon as we have these functions, ie. sooner in the sequence of commits and potentially in the commit that adds this method (can squash later, but keep it adjacent).

neiljp reviewed (commented):
@Subhasish-Behera This looks much different - and more detailed - and certainly seems to work in practice :+1:

Your commit style is broadly OK, but there are a few elements that are throwing the style off:
- there's no blank line between the title and summary (so tig is showing it all on one line for me!)
- it's challenging, but it works best, including for github breaking the title into multiple lines, if you can keep the title to 72 characters
- using gitlint as per the README should help prompt for these issues; if not then we can adjust the rules it uses perhaps
- for referring to future work, it's useful to refer to the principle of why you're doing it, rather than eg. specific function names that don't yet exist :)

The tests are generally good to see, with a few points to bear in mind:
- if adding a test for existing code, do it before modifying; you can then be more sure that you don't break anything
- if adding a function, generally add the test in that commit too
- it's not always necessary to test every underlying helper function
- similarly, it's good to aim for the big picture behavior, though sometimes it can be challenging not to test the implementation

I like how you extracted the function to handle the common code, in addition to the one for the base PM/DM setup, though I do wonder if it would be simpler to incorporate it into the existing similar function.

I'll review the tests in more detail later, since we may wish to restructure again slightly before we're done, as per my comments - this is another reason that testing every little helper function can be a challenge, since it's testing the implementation in some sense!

Commit:
boxes: Footer message if tab key pessed while editing a private message.

Commit:
boxes: Disable cycling while editing a Private Message.

The WriteBox has same keypress for the box created for editing as well
as reply. There is no neet to check the validity of self.to_write_box in
case of editing as it's non-editable.So now it checks the msg_edit_state
. Also now if the focus_position is FOCUS_CONTAINER_MESSAGE then the tab
will change the focus to FOCUS_CONTAINER_HEADER only if its a reply
(msg_edit_state =None).
Tests Adapted.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -172,41 +172,11 @@ def send_stop_typing_status(self) -> None:
-            recipient_info = ""
-
-        self.send_next_typing_update = datetime.now()
-        self.to_write_box = ReadlineEdit("To: ", edit_text=recipient_info)
-        self.to_write_box.enable_autocomplete(
-            func=self._to_box_autocomplete,
-            key=primary_key_for_command("AUTOCOMPLETE"),
-            key_reverse=primary_key_for_command("AUTOCOMPLETE_REVERSE"),
-        )
-        self.to_write_box.set_completer_delims("")
```
If we shouldn't be sending typing updates for editing, I suspect that could be clearer in the API documentation. If you're sure this is the behavior we want, posting to that stream that you think that's the case would be useful, or even a quick PR for it.

Since we're splitting the private-compose and private-edit UI it certainly makes it simpler to separate this behavior in this way.

neiljp commented:
@Subhasish-Behera Is this ready for another review?

Subhasish-Behera committed:
refactor: Change name of update_recipients function.

Change the name of update_recipients to
update_recipients_from_emails_in_widget.

Subhasish-Behera commented:
> @Subhasish-Behera Is this ready for another review?

Yep.

neiljp reviewed (commented):
@Subhasish-Behera This looks like it will be really useful! :+1:

I did leave a lot of comments, but some relate to commit structure, commit text, and smaller changes or suggestions.

It seems like the functionality is essentially done, so this should be a case of moving things around and some final tidying, with some optional extensions.

Other than fixing the issue fully, this will be a good refactor to the compose code to have in place for future work! :tada:
